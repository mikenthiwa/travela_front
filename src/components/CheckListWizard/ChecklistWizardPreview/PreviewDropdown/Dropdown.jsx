import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { isEqual } from 'lodash';
import './index.scss';
import selectDropdownIcon from '../../../../images/icons/form_select_dropdown.svg';

class Dropdown extends Component {
  state = {
    open: false,
    option: {}
  };
  container = React.createRef();
  input = React.createRef();

  componentDidMount() {
    document.addEventListener('click', this.handleClickOutside);
  }

  componentDidUpdate() {
    const { configuration: { options } } = this.props;
    const { option } = this.state;
    const newOption =  options.find(item => item.id === option.id) || {};
    if (!isEqual(option, newOption)) return this.handleItemChoice(newOption);
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.handleClickOutside);
  }

  handleClickOutside = (event) => {
    if (this.container.current && !this.container.current.contains(event.target)) {
      this.setState({
        open: false
      });
    }
  };

  handleClick = () => {
    this.setState((state) => {
      return {
        open: !state.open
      };
    });
  };

  handleItemChoice = (option) => {
    const { handleCheckName } = this.props;
    this.setState({ option, open: false }, handleCheckName(option.behaviour));
  };

  render() {
    const { configuration } = this.props;
    const { options } = configuration;
    const { open, option } = this.state;
    return (
      <div>
        <div ref={this.container} className="style-container">
          <input
            className="input"
            type="text"
            onClick={this.handleClick}
            value={option.name || ''}
            readOnly="readOnly"
          />
          <img 
            src={selectDropdownIcon} 
            alt="icn" 
            onClick={this.handleClick} 
            role="presentation" 
          />
          {open && (
            <div className="dropdown">
              {options.map((option) => (
                <ul 
                  key={option.id}
                  role="presentation"
                  onClick={() => this.handleItemChoice(option)}
                > 
                  <li>
                    {option.name}
                  </li>
                </ul>
              ))}
            </div>
          )}
        </div>
      </div>
    );
  }
}

Dropdown.propTypes = {
  handleCheckName: PropTypes.func.isRequired,
  configuration: PropTypes.shape({options: PropTypes.array}).isRequired,
};

export default Dropdown;
