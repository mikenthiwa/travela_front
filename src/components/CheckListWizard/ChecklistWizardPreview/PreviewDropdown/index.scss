.style-container {
  position: relative;
  display: inline-block;

  img {
    position: absolute;
    top: 17px;
    right: 15px;
    transition: transform 0.05s ease-out;
    transform-origin: center;
  }
  
  .input {
    width: 300px;
    padding: 10px 10px 10px 15px;
    border: 1px solid #c3c3c3;
    font-size: 14px;
    cursor: pointer;
  }
  
  .dropdown {
    position: absolute;
    border: 1px solid #c3c3c3;
    max-height: 200px;
    height: fit-content;
    background: white;
    box-shadow: 2px 0 5px 2px rgba(0, 0, 0, .1);
    border-radius: 3px;
    margin-top: 5px;
    z-index: 15;
    width: 100%;
    overflow: hidden;
    overflow-y: scroll;
    li {
      padding-left: 15px;
      margin: opx;
      width: 100%;
      height: 35px;
      cursor: pointer;
      display: flex;
      align-items: center;
      justify-items: center;
      transition: 0.4s all;
      &:hover{
        background-color: #F0F3F7;
        color: #3d45e3;
      }
    }
    ul {
      padding-inline-start: 0px;
    }
  }
}
