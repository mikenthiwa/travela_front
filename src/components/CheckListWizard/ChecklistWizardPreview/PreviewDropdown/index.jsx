import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import Dropdown from './Dropdown';
import PreviewBehaviour from '../PreviewBehaviour';
import './index.scss';

class PreviewDropdown extends Component {
  state = {
    behaviour: {},
    showBehaviour: false
  }
  handleCheckName = (behaviour) => {
    this.setState({ behaviour, showBehaviour: !!behaviour});
  }

  render() {
    const { item: { configuration, prompt } } = this.props;
    const { behaviour, showBehaviour } = this.state;
    
    return (
      <div>
        <div>
          <div>
            <Dropdown   
              prompt={prompt} 
              configuration={configuration} 
              handleCheckName={this.handleCheckName}
            />
          </div>
          {showBehaviour && (
            <PreviewBehaviour 
              behaviour={behaviour || {}} 
            />)}
        </div>
      </div>
    );
  }
}

PreviewDropdown.propTypes = {
  item: PropTypes.object.isRequired
};

export default PreviewDropdown;
