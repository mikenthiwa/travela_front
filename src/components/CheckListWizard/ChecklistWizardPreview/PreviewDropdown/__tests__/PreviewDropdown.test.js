import React from 'react';
import { shallow, mount} from 'enzyme';
import PreviewDropdown from '..';

const props = {
  item: {
    prompt: 'Do you have a valid visa?',
    order: 1,
    configuration: {
      options: [
        {
          id: 1,
          name: 'Apple',
          behaviour: {
            action: 'preview dropdown',
            action: {
              type: '',
              payload: ''
            }
          }
        }
      ]
    },
  },
  behaviour: {},
  showBehaviour: false,
  handleCheckName: jest.fn()
};

describe('<PreviewDropdown />', () => {
  it('should render without crashing', () => {
    const wrapper = mount(<PreviewDropdown {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle check name', () => {
    const wrapper = mount(<PreviewDropdown {...props} />);
    const instance = wrapper.instance();
    wrapper.setState({
      behaviour: {},
      showBehaviour: true,
      payload: ''
    });
    instance.handleCheckName(props.behaviour);
    expect(instance.state.showBehaviour).toBe(true);
  });
});
