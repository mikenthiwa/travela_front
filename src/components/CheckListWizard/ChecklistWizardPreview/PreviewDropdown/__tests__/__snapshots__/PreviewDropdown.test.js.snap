// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PreviewDropdown /> should render without crashing 1`] = `
<div>
  <div>
    <div>
      <div>
        <div
          className="style-container"
        >
          <input
            className="input"
            onClick={[Function]}
            readOnly="readOnly"
            type="text"
            value=""
          />
          <img
            alt="icn"
            onClick={[Function]}
            role="presentation"
            src="form_select_dropdown.svg"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
